src/shape_extraction.cpp

```diff
@@ -226,40 +227,91 @@ bool FitBox(const PointCloudC::Ptr& input,
-      pose->orientation.x = q.x();
-      pose->orientation.y = q.y();
-      pose->orientation.z = q.z();
-      pose->orientation.w = q.w();
+      last_x_max = x_max;
+      last_x_min = x_min;
+      last_y_max = y_max;
+      last_y_min = y_min;
 
       min_volume = area * height;
     }
   }
 
+  
+  double x_dim = last_x_max - last_x_min;
+  double y_dim = last_y_max - last_y_min;
+
+  Eigen::Matrix3f adjusted_transformation = Eigen::Matrix3f::Zero();
+  MakeGoodBoxOrientation(model, x_dim, y_dim, transformation, &adjusted_transformation);
+
+  if (x_dim > y_dim) {
+    dimensions->x = (last_y_max - last_y_min);
+    dimensions->y = (last_x_max - last_x_min);
+  } else {
+    dimensions->x = (last_x_max - last_x_min);
+    dimensions->y = (last_y_max - last_y_min);
+  }
+  dimensions->z = height;
+
+  Eigen::Quaternionf q(adjusted_transformation);
+  pose->orientation.x = q.x();
+  pose->orientation.y = q.y();
+  pose->orientation.z = q.z();
+  pose->orientation.w = q.w();
+
   return true;
 }
+
+void MakeGoodBoxOrientation(const pcl::ModelCoefficients::Ptr& plane_coeff,
+		double x_dimension,
+		double y_dimension,
+		const Eigen::Matrix3f& rotation_matrix,
+		Eigen::Matrix3f* output_matrix) {
+  // Flip orientation if necessary to force x dimension < y dimension
+  if (x_dimension > y_dimension) {
+    Eigen::Vector3f y_axis = rotation_matrix.col(1);
+    // There are two choices for the new x axis. This chooses the one that
+    // is closer to the positive x direction of the data.
+    if (y_axis.x() < 0) {
+      y_axis = -1 * rotation_matrix.col(1);
+    }
+    output_matrix->col(0) = y_axis;
+    output_matrix->col(1) = rotation_matrix.col(2).cross(y_axis);
+  } else {
+    output_matrix->col(0) = rotation_matrix.col(0);
+    output_matrix->col(1) = rotation_matrix.col(1);
+  }
+  output_matrix->col(2) = rotation_matrix.col(2);
+
+  // Compute the approximated expected vectors of box
+  Eigen::Vector3f x_axis(1.0, 0.0, 0.0);
+  Eigen::Vector3f z_axis(plane_coeff->values[0], plane_coeff->values[1],
+                         plane_coeff->values[2]);
+  Eigen::Vector3f y_axis = z_axis.cross(x_axis);
+
+  // Check if the object is facing towards or perpendicular to the positive
+  // x-axis. If not, the angle theta between x basis vector and x axis should be
+  // 90 < theta <= 180, which means the result of dot product of the two
+  // vectors would be negative.
+  if (output_matrix->col(0).dot(x_axis) < 0.0) {
+    ROS_WARN(
+        "The box doesn't face towards the positive x-axis. It has x basis "
+        "vector (%f, %f, %f) with dot product result of %f",
+        output_matrix->col(0)(0), output_matrix->col(0)(1),
+        output_matrix->col(0)(2), output_matrix->col(0).dot(x_axis));
+    output_matrix->col(0) = output_matrix->col(0) * -1.0;
+  }
+
+  // Check if the object has the same z-axis as the normal vector of the plane.
+  if ((output_matrix->col(2) - z_axis).array().abs().matrix().sum() > 0.0001) {
```

Comment:
If this doesn't ever happen, then just delete the whole if statement. In my earlier comments, I mentioned that you probably don't need to pass in the ModelCoefficients at all. It looks like this is just comparing the ModelCoefficients to the z axis in the rotation matrix, which should already be the same.

Reply:
Completed